llvm/lib/Target/AMDGPU/SIRegisterInfo.cpp

```diff
@@ -693,20 +699,29 @@ BitVector SIRegisterInfo::getReservedRegs(const MachineFunction &MF) const {
     }
   }
 
-  for (unsigned i = MaxNumVGPRs; i < TotalNumVGPRs; ++i) {
-    unsigned Reg = AMDGPU::VGPR_32RegClass.getRegister(i);
-    reserveRegisterTuples(Reserved, Reg);
+  for (const TargetRegisterClass *RC : regclasses()) {
+    if (RC->isBaseClass() && isVGPRClass(RC)) {
+      unsigned NumRegs = divideCeil(getRegSizeInBits(*RC), 32);
+      for (MCPhysReg Reg : *RC) {
```

Comment:
drop_front by NumRegs probably works?

Reply:
It would need adjusting for the Aligned classes. And in general I wasn't sure whether iterating RC is guaranteed to give you registers in order, and does that guarantee that HWRegIndex will also be in order?

Reply:
I don't know if it's formally guaranteed, but with the way it's generated it always will be in order

Reply:
It's also tricky to get this right for 16-bit classes. I'd prefer to leave it to a follow-up.